### PR TITLE
Refactor code to use KVListWithOptions to get keys

### LINKED ISSUE #441: Refactor code to use KVListWithOptions to get keys
New helper functionality was added in https://github.com/mattermost/mattermost-server/issues/12936 to make writing plugin code that performs operations on a set of keys easier to write.  

At this time there are several places in the jira plugin where this happens:
* `CountUsers`
* `debugResetStats`
* `consolidatedStoredStats`

Update these areas to use the new `KVListWithOptions` helper function and remove the `listPerPage` var.

### PR DESCRIPTION
**Pull request by AGMETEOR**
Fixes #441

### CODE REVIEW

**Comment by levb** [comment]
In order to build and merge this PR, we'd need to advance the required server version to (5.18? 5.20?) which seems premature. I wonder if we can keep track of such changes by having a feature branch that tracks master, but includes changes requiring the more recent server.

At some point, we may decide to merge (or cherry-pick) some of them into master, and bump the requirement.

@mickmister @jfrerich @hanzei @cpoile thoughts?

**Comment by cpoile** [comment]
@hanzei is `KVListWithOptions` working? Or am I thinking of the problems we had with `KVSetWithOptions`?
As for a parallel branch -- seems complicated. If we can't upgrade to 5.18 or 5.20 because some customer's can't, then we should 1/5 hold off on this.

**Comment by levb** [comment]
`KVListWithOptions` was just merged to master, so 5.20 is my guess. This is not not the first nor the only time that we run into the issue. Maybe then we have a special label on these PRs that would tell us they are not "stale". Or use an existing Do Not Merge/Awaiting... one?

**Comment by hanzei** [comment]
Just to be clear: `KVListWithOptions` is helper method and requires MM 5.6 to use. This is documented in the godoc https://github.com/mattermost/mattermost-server/pull/13576/files#diff-828165bc3b7f7ef392a24252cf6e0f97R61

Hence, in order to use it, a bump of the dependency `github.com/mattermost/mattermost-server` to the commit that added the helper is needed. There is no need to bump the required server version.

**Comment by levb** [comment]
> Just to be clear: `KVListWithOptions` is helper method and requires MM 5.6 to use.

Right. The gotcha that newer Helper methods are still binary-compatible with older servers continues to confuse me. Sorry about the misunderstanding.

> Hence, in order to use it, a bump of the dependency `github.com/mattermost/mattermost-server` to the commit that added the helper is needed. There is no need to bump the required server version.

Right. We still should probably complete the v5 migration first in the plugin, then it'll be cleaner.

@jfrerich @aaronrothschild @mickmister I am thinking, since 2.4 is such an incremental release, maybe we should use the chance and update the server version requirement. 2.3.2(+) can be our supported version for the older servers. Or call it 3.0 for that reason? Will move this conversation to the Jira channel.

**Comment by hanzei** [comment]
> Right. We still should probably complete the v5 migration first in the plugin, then it'll be cleaner.

Isn't this one for all plugins?

File: server/kv.go [hunk 1]
```diff
@@ -484,42 +485,47 @@ func (store store) DeleteUserInfo(ji Instance, mattermostUserId string) (returnE
...
-		if appErr != nil {
-			return 0, appErr
-		}
 
-		for _, key := range keys {
-			// User records are not currently prefixed. Consider any 32-hex key.
-			if !reHexKeyFormat.MatchString(key) {
-				continue
-			}
+	options := []plugin.KVListOption{
```
Comment (levb) [inline server/kv.go]: 1/5 inline the func() here

**Comment by jasonblais** [comment]
Hi @AGMETEOR, touching base to see if you have any questions about the feedback above? We'd be happy to clarify :

**Comment by AGMETEOR** [comment]
> Hi @AGMETEOR, touching base to see if you have any questions about the feedback above? We'd be happy to clarify 

@jasonblais looks like the PR this depends on is not yet merged so i had paused work on this.

**Comment by hanzei** [comment]
@AGMETEOR I think every needed change for the server has been merged (https://github.com/mattermost/mattermost-server/pull/13576) and you are good to go to continue with your work

**Comment by AGMETEOR** [comment]
> @AGMETEOR I think every needed change for the server has been merged ([mattermost/mattermost-server#13576](https://github.com/mattermost/mattermost-server/pull/13576)) and you are good to go to continue with your work

Hello @hanzei , running `go get github.com/mattermost/mattermost-server/v5` adds `v5.20.1` but I don't seem to find the helper function `KVListWithOptions`

**Comment by hanzei** [comment]
@AGMETEOR You need to pull the changes from `master`. Because https://github.com/mattermost/mattermost-server/commit/6377bfef35bcc755e17819a1567063687f0a31be added this helper the correct command is `go get github.com/mattermost/mattermost-server/v5@6377bfef35bcc755e17819a1567063687f0a31be`

**Comment by AGMETEOR** [comment]
> @AGMETEOR You need to pull the changes from `master`. Because [mattermost/mattermost-server@6377bfe](https://github.com/mattermost/mattermost-server/commit/6377bfef35bcc755e17819a1567063687f0a31be) added this helper the correct command is `go get github.com/mattermost/mattermost-server/v5@6377bfef35bcc755e17819a1567063687f0a31be`

Thanks Ben.

**Comment by AGMETEOR** [comment]
> @AGMETEOR You need to pull the changes from `master`. Because [mattermost/mattermost-server@6377bfe](https://github.com/mattermost/mattermost-server/commit/6377bfef35bcc755e17819a1567063687f0a31be) added this helper the correct command is `go get github.com/mattermost/mattermost-server/v5@6377bfef35bcc755e17819a1567063687f0a31be`

@hanzei I am facing issues running `go get github.com/mattermost/mattermost-server/v5@6377bfef35bcc755e17819a1567063687f0a31be`

* First issue is that somehow the package keeps snapping back to v5.19.0
* Then even it's on version `github.com/mattermost/mattermost-server/v5 v5.3.2-0.20200217204434-6377bfef35bc` the following errors occur
```
could not import github.com/mattermost/mattermost-plugin-autolink/server/autolinkclient (no package for import github.com/mattermost/mattermost-plugin-autolink/server/autolinkclient)

could not import github.com/mattermost/mattermost-plugin-workflow-client/workflowclient (no package for import github.com/mattermost/mattermost-plugin-workflow-client/workflowclient)

store.plugin undefined (type store has no method plugin)

```
Any ideas how to get this working?

**Comment by hanzei** [comment]
@AGMETEOR I'm running into the same problem and don't know know to fix it. It keeps snapping back to `5.19.0` :man_shrugging: 

Would you mind waiting until RC-1 of MM 5.22 is cut, which should be in roughly 3 weeks.

cc @levb Maybe you have ideas who how to fix this

**Comment by AGMETEOR** [comment]
> @AGMETEOR I'm running into the same problem and don't know know to fix it. It keeps snapping back to `5.19.0` 
> 
> Would you mind waiting until RC-1 of MM 5.22 is cut, which should be in roughly 3 weeks.
> 
> cc @levb Maybe you have ideas who how to fix this

@hanzei thanks for trying it out as well. One thing I did was to upgrade and then immediately disconnect my internet. That way it didn't snap back. Otherwise, I was starting to think maybe my computer and in particular git was having a problem.

**Comment by hanzei** [comment]
@AGMETEOR I went ahead and fixed the dependency issues. Now only test fail. Please let me know if you need help fixing them.

**Comment by hanzei** [comment]
@AGMETEOR I fixed the tests. I hope you don't mind.

@jfrerich @levb This is ready for another review.

**Comment by AGMETEOR** [comment]
> @AGMETEOR I fixed the tests. I hope you don't mind.
> 
> @jfrerich @levb This is ready for another review.

Hello @hanzei , thanks for working on this. It's been difficult for me to find time to work on all the issues I left pending because of work. I am sorry for any delays.

**Comment by levb** [comment]
@DHaussermann the user count in stats should be checked

**Comment by levb** [comment]
Thanks @AGMETEOR! Thanks for continuing your contributions!

**Comment by DHaussermann** [comment]
Hi @AGMETEOR  I tested this out connected to our Cloud instance and saw some strange behavior.  

1. Until I reset the stats both commands to pull them up returned `Key has no prefix`
2. Even when generating fresh stats - `Key is not 32-hex`
3. I tried to uninstall and reinstall my Jira instance to see if it would resolve the key issue and see a different issue now. 

It looks like testing is blocked by a bug resolved with https://github.com/mattermost/mattermost-plugin-jira/pull/534 Is it possible to pull in from master to unblock testing? 
cc @hanzei

**Comment by hanzei** [comment]
@DHaussermann Done

**Comment by DHaussermann** [comment]
@AGMETEOR  After the merge from master, the install issue has been resolved. 
After I reinstall I still see the issues related to the key so, I can't confirm connected users is working. 

1. `/jira stats jira` shows `Key has no prefix`
2. `/jira debug stats expvar jira` shows `Key is not 32-hex`
StatsKeyIssue

Is it possible jira debug needs additional changes? I double checked the master branch and the stats feature seems to work as expected without showing issues related to the key.

**Comment by jasonblais** [comment]
Hey @AGMETEOR - I know it has been a while. Are you interested continuing to work on this pull request? Looks like it's close. cc @jfrerich

**Comment by jfrerich** [comment]
Hello @AGMETEOR! I hope you have been well!  If you are not interested in this ticket, would you mind if one of our developers pick it up from here so we can merge your changes?  Thanks!

**Comment by AGMETEOR** [comment]
> Hello @AGMETEOR! I hope you have been well! If you are not interested in this ticket, would you mind if one of our developers pick it up from here so we can merge your changes? Thanks!

Hello @jfrerich , everything is well. I hope everything is well with you too. Unfortunately, I am unable to work on this anymore. Yes, one of your developers can pick it :)

**Comment by jfrerich** [comment]
@AGMETEOR, thank you so much for letting me know and we'll pick this up from here.  Thanks again for all the great contributions over the last year and aI wish you all the best!

**Comment by hanzei** [comment]
Closing as inactive

**Pull request closed without merging**